# User ratings EDA
We found a data set containing the BoardGameGeeks user ratings as of May2019 on Kaggle: [boardgame ratings on kaggle](https://www.kaggle.com/datasets/thedevastator/board-game-ratings-by-country)  

The data archive contains the following csv files:
* 2019-05-02.csv
* bggreviews_small.csv
* games_detailed_info.xlsx
* games_detailed_info_filtered.csv
* popular_board_game.csv
* users.csv

In [30]:
# import setup
from setup import *

In [31]:
# function for reading the dataset and displaying some basic information
def read_and_explore(file , df_name, path= '../data/'):
    dataframe = pd.read_csv(path+file)
    print('The dataframe', df_name,'has', dataframe.shape[0], 'rows and',  dataframe.shape[1],'columns')
    for column in dataframe.columns:
        print('Column', column,'has', dataframe[column].nunique(), 'unique values')
    print('Missing values in columns:')
    print(dataframe.isna().sum())
    return dataframe

File 2019-05-02.csv contains information (ID, Name, Year, Rank, Average, Bayes average, Users rated, URL, and Thumbnail) of 17065 games. 
Not needed.

In [32]:
df_2019 = read_and_explore(file='2019-05-02.csv', df_name='df_2019')
df_2019.head()


The dataframe df_2019 has 17065 rows and 9 columns
Column ID has 17065 unique values
Column Name has 16735 unique values
Column Year has 182 unique values
Column Rank has 17065 unique values
Column Average has 602 unique values
Column Bayes average has 1872 unique values
Column Users rated has 2622 unique values
Column URL has 17065 unique values
Column Thumbnail has 17016 unique values
Missing values in columns:
ID                0
Name              0
Year              0
Rank              0
Average           0
Bayes average     0
Users rated       0
URL               0
Thumbnail        45
dtype: int64


ID                                           Name  Year  Rank  Average   
0  174430                                     Gloomhaven  2017     1     8.90  \
1  161936                      Pandemic Legacy: Season 1  2015     2     8.64   
2  182028  Through the Ages: A New Story of Civilization  2015     3     8.53   
3  167791                              Terraforming Mars  2016     4     8.40   
4   12333                              Twilight Struggle  2005     5     8.33   

   Bayes average  Users rated   
0          8.611        24574  \
1          8.489        30921   
2          8.258        15551   
3          8.242        38331   
4          8.175        34268   

                                                 URL   
0                       /boardgame/174430/gloomhaven  \
1         /boardgame/161936/pandemic-legacy-season-1   
2  /boardgame/182028/through-ages-new-story-civil...   
3                /boardgame/167791/terraforming-mars   
4                 /boardgame/12333/twilight-struggle   

                                           Thumbnail  
0  https://cf.geekdo-images.com/micro/img/8JYMPXd...  
1  https://cf.geekdo-images.com/micro/img/KbAKyhb...  
2  https://cf.geekdo-images.com/micro/img/APvZ_BY...  
3  https://cf.geekdo-images.com/micro/img/71v8jF5...  
4  https://cf.geekdo-images.com/micro/img/uStZr7_...

File users.csv contains information (users, url, country) of 289914 users. 
Country could be used for filtering to reduce number of users and ratings

In [33]:
df_users = read_and_explore(file='users.csv', df_name='df_users')
# print(df_users['country'].unique())
df_users.head()


The dataframe df_users has 289914 rows and 3 columns
Column users has 289913 unique values
Column url has 289914 unique values
Column country has 226 unique values
Missing values in columns:
users          1
url            0
country    72873
dtype: int64


users                                         url        country
0   sidehacker   https://boardgamegeek.com/user/sidehacker  United States
1  Varthlokkur  https://boardgamegeek.com/user/Varthlokkur  United States
2   dougthonus   https://boardgamegeek.com/user/dougthonus  United States
3       cypar7       https://boardgamegeek.com/user/cypar7  United States
4      ssmooth      https://boardgamegeek.com/user/ssmooth  United States

File bggreviews_small.csv contains 13 million ratings for 17.000 games from 290.000 users. This is the file we need.  
66 rows with missing user need to be dropped.

In [34]:
# read data from csv files
df_bgg_small = read_and_explore(file='bggreviews_small.csv', df_name='df_bgg_small')
df_bgg_small.head()

The dataframe df_bgg_small has 13170073 rows and 3 columns
Column user has 289913 unique values
Column rating has 8202 unique values
Column ID has 17065 unique values
Missing values in columns:
user      66
rating     0
ID         0
dtype: int64


user  rating  ID
0   sidehacker    10.0  13
1  Varthlokkur    10.0  13
2   dougthonus    10.0  13
3       cypar7    10.0  13
4      ssmooth    10.0  13

In [35]:
#rename: ID = game_id
df_user_ratings_ = df_bgg_small.copy()
df_user_ratings_.rename(columns={'ID':'game_id'}, inplace=True)
df_user_ratings_.head()

user  rating  game_id
0   sidehacker    10.0       13
1  Varthlokkur    10.0       13
2   dougthonus    10.0       13
3       cypar7    10.0       13
4      ssmooth    10.0       13

In [36]:
# check missing values
missing_user = df_user_ratings_[df_user_ratings_.isna().any(axis=1)]
missing_user.head()


user  rating  game_id
63824   NaN     6.0       13
164185  NaN     5.0      822
383996  NaN     5.0    36218
447312  NaN     4.0     9209
547228  NaN     7.0     3076

In [37]:
# drop rows with missing username
df_user_ratings = df_user_ratings_.dropna().reset_index(drop=True)
print('The dataframe has', df_user_ratings.shape[0], 'rows and',  df_user_ratings.shape[1],'columns')
print('Missing values in columns:')
print(df_user_ratings.isna().sum())


The dataframe has 13170007 rows and 3 columns
Missing values in columns:
user       0
rating     0
game_id    0
dtype: int64


In [38]:
df_user_ratings.to_csv('../data/user_ratings.csv')

### Filter user ratings to keep only those for games that are ranked > 0

In [ ]:
# import the data with filter on rank >0
df_names = pd.read_csv('../data/boardgames_ranks.csv').query('rank > 0')
df_names.shape

(27925, 16)

In [48]:
# keep only ratings for games that are in the 30000 selected 
user_ratings_small = df_user_ratings[df_user_ratings['game_id'].isin(df_names['id'])].reset_index(drop=True)
print(user_ratings_small.shape)
user_ratings_small.tail()

(13162663, 3)


user  rating  game_id
13162658       pswissler     3.0     8256
13162659   Sprayoncrayon     3.0     8256
13162660     PhoenixSong     2.5     8256
13162661         tsantos     2.0     8256
13162662  Mighty Florist     2.0     8256

In [49]:
# new dataset
print('The dataframe has', user_ratings_small.shape[0], 'rows and',  user_ratings_small.shape[1],'columns')
for column in user_ratings_small.columns:
    print('Column', column,'has', user_ratings_small[column].nunique(), 'unique values')
print('Missing values in columns:')
print(user_ratings_small.isna().sum())

The dataframe has 13162663 rows and 3 columns
Column user has 289870 unique values
Column rating has 8196 unique values
Column game_id has 16999 unique values
Missing values in columns:
user       0
rating     0
game_id    0
dtype: int64


In [50]:
# old dataset for comparison
print('The dataframe has', df_user_ratings.shape[0], 'rows and',  df_user_ratings.shape[1],'columns')
for column in df_user_ratings.columns:
    print('Column', column,'has', df_user_ratings[column].nunique(), 'unique values')
print('Missing values in columns:')
print(df_user_ratings.isna().sum())

The dataframe has 13170007 rows and 3 columns
Column user has 289913 unique values
Column rating has 8202 unique values
Column game_id has 17065 unique values
Missing values in columns:
user       0
rating     0
game_id    0
dtype: int64


__old:__  
The dataframe has 13170007 rows and 3 columns  
Column user has 289913 unique values  
Column rating has 8202 unique values  
Column game_id has 17065 unique values  
Missing values in columns:
user       0
rating     0
game_id    0
dtype: int64

__new:__ 
The dataframe has 13162663 rows and 3 columns  
Column user has 289870 unique values  
Column rating has 8196 unique values  
Column game_id has 16999 unique values  
Missing values in columns:
user       0
rating     0
game_id    0
dtype: int64  

In [ ]:
# write to user_ratings_small to csv
user_ratings_small.to_csv('../data/user_ratings_small.csv', index= False)